File: dsa/concepts/03_slidingwindow/java/L239_SlidingWindowMaximum.java
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

public class L239_SlidingWindowMaximum {

	public static void main(String[] args) {
		L239_SlidingWindowMaximum obj = new L239_SlidingWindowMaximum();

		int[] nums = new int[] {1,3,-1,-3,5,3,6,7};
		int k = 3;
		int[] result = obj.maxSlidingWindow(nums, k);
		System.out.println(Arrays.toString(result));
	}

	// Problem: https://leetcode.com/problems/sliding-window-maximum/
	// Idea: Use, monotonically decreasing queue. Maintain the nums in decreasing order in queue.
	// 		first<-	5, 4, 3  <- last
	// Take the left most to take the max for that window.
	// Remove the left most if it goes out of window.
	public int[] maxSlidingWindow(int[] nums, int k) {
		int n = nums.length;
		int[] ans = new int[n-k+1];

		int l=0, r=0;
		Deque<Integer> dq = new ArrayDeque<>();				// stores indices
		while(r<n) {
			while(!dq.isEmpty() && nums[dq.peekLast()] < nums[r])
				dq.removeLast();

			dq.add(r);					// add and remove one element. window size will be maintained.
			if(dq.peek() < l) {
				dq.remove();
			}

			if(r+1 >= k) {				// this condition to skip first indices like 0,1 etc.
				ans[l++] = nums[dq.peek()];
			}
			r++;
		}

		return ans;
	}

}

File: dsa/concepts/03_slidingwindow/python/l239_sliding_window_maximum.py
"""
L239 - Sliding Window Maximum

Problem: https://leetcode.com/problems/sliding-window-maximum/
Idea: Use monotonically decreasing deque. Maintain the nums in decreasing order in deque.
first <- 5, 4, 3 <- last
Take the left most to take the max for that window.
Remove the left most if it goes out of window.
Time: O(n)
Space: O(k)
"""

from collections import deque

def max_sliding_window(nums, k):
    n = len(nums)
    ans = [0] * (n - k + 1)
    
    l, r = 0, 0
    dq = deque()  # stores indices
    
    while r < n:
        # Remove elements from back while current element is larger
        while dq and nums[dq[-1]] < nums[r]:
            dq.pop()
        
        dq.append(r)  # add current element index
        
        # Remove elements that are out of current window
        if dq[0] < l:
            dq.popleft()
        
        # If we have processed at least k elements
        if r + 1 >= k:
            ans[l] = nums[dq[0]]  # front of deque has max element
            l += 1
        
        r += 1
    
    return ans


if __name__ == "__main__":
    nums = [1, 3, -1, -3, 5, 3, 6, 7]
    k = 3
    print(f"Input: nums = {nums}, k = {k}")
    print(f"Output: {max_sliding_window(nums, k)}")
